src/counter.rs

```diff
@@ -0,0 +1,101 @@
+// Copyright 2014 The Prometheus Authors
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::{Arc, RwLock};
+
+use proto;
+use metrics::{Metric, Opts};
+use value::{Value, ValueType};
+use desc::Desc;
+use errors::Result;
+
+/// `Counter` is a Metric that represents a single numerical value that only ever
+/// goes up.
+#[derive(Clone)]
+pub struct Counter {
+    desc: Arc<Desc>,
+    v: Arc<RwLock<Value>>,
+}
+
+impl Counter {
+    pub fn new<S: Into<String>>(namespace: S, sub_system: S, name: S, help: S) -> Result<Counter> {
+        let opts = Opts::new(namespace, sub_system, name, help);
+        Counter::with_opts(opts)
+    }
+
+    pub fn with_opts(opts: Opts) -> Result<Counter> {
+        let desc = try!(Desc::new(opts.fq_name(), opts.help, vec![], opts.const_labels));
+        let v = try!(Value::new(&desc, ValueType::Counter, 0.0, vec![]));
+        Ok(Counter {
+            desc: Arc::new(desc),
+            v: Arc::new(RwLock::new(v)),
+        })
+    }
+}
+
+impl Counter {
+    /// `add` adds the given value to the counter. It panics if the value is <
+    /// 0.
+    #[inline]
+    pub fn add(&mut self, v: f64) {
+        assert!(v >= 0.0, "counter cannot decrease in value");
```

Comment:
As a library user, I prefer knowing about bad inputs by getting back an `Error`, instead of having my application blowing up on a library. Also, it may be interesting knowing what is the new resulting value in an atomic way. What about returning a `Result<f64, std::num::ParseFloatError>` (or something similar) instead?